src/server/frontend_xwayland/xwayland_server.cpp

```diff
@@ -38,93 +32,44 @@
 #include <unistd.h>
 
 #include <chrono>
-#include <thread>
 #include <condition_variable>
 
 namespace mf = mir::frontend;
 namespace md = mir::dispatch;
 using namespace std::chrono_literals;
 
-mf::XWaylandServer::XWaylandServer(
-    std::shared_ptr<WaylandConnector> const& wayland_connector,
-    XWaylandSpawner const& spawner,
-    std::string const& xwayland_path)
-    : wayland_connector{wayland_connector},
-      xwayland_path{xwayland_path}
+namespace
 {
-    enum { server, client, size };
-    int wl_client_fd[size], wm_fd[size];
-
-    if (socketpair(AF_UNIX, SOCK_STREAM | SOCK_CLOEXEC, 0, wl_client_fd) < 0)
-    {
-        // "Shouldn't happen" but continuing is weird.
-        mir::fatal_error("wl connection socketpair failed");
-        return; // doesn't reach here
-    }
-
-    if (socketpair(AF_UNIX, SOCK_STREAM | SOCK_CLOEXEC, 0, wm_fd) < 0)
-    {
-        // "Shouldn't happen" but continuing is weird.
-        mir::fatal_error("wm fd socketpair failed");
-        return; // doesn't reach here
-    }
-
-    mir::log_info("Starting Xwayland");
-    xwayland_pid = fork();
-
-    switch (xwayland_pid)
-    {
-    case -1:
-        mir::fatal_error("Failed to fork");
-        return; // Keep compiler happy (doesn't reach here)
-
-    case 0:
-        close(wl_client_fd[server]);
-        close(wm_fd[server]);
-        execl_xwayland(spawner, wl_client_fd[client], wm_fd[client]);
-        return; // Keep compiler happy (doesn't reach here)
-
-    default:
-        close(wl_client_fd[client]);
-        close(wm_fd[client]);
-        x11_fd = Fd{wm_fd[server]};
-        wayland_fd = Fd{wl_client_fd[server]};
-        connect_wm_to_xwayland();
-        break;
-    }
-}
-
-mf::XWaylandServer::~XWaylandServer()
+/// client and server are symetrical, they only differ in how they are used
+struct SocketPair
 {
-    mir::log_info("Deiniting xwayland server");
+    mir::Fd client;
+    mir::Fd server;
+};
 
-    // Terminate any running xservers
-    if (kill(xwayland_pid, SIGTERM) == 0)
+auto make_socket_pair() -> SocketPair
+{
+    int pipe[2];
+    if (socketpair(AF_UNIX, SOCK_STREAM | SOCK_CLOEXEC, 0, pipe) < 0)
     {
-        std::this_thread::sleep_for(100ms);// After 100ms...
-        if (kill(xwayland_pid, 0) == 0)    // ...if Xwayland is still running...
-        {
-            mir::log_info("Xwayland didn't close, killing it");
-            kill(xwayland_pid, SIGKILL);     // ...then kill it!
-        }
+        BOOST_THROW_EXCEPTION(std::system_error(errno, std::generic_category(), "Creating socket pair failed"));
```

Comment:
Extremely niche nit: this should probably be `std::system_category` (as it's a OS-generated error). It's functionally the same, though, as we don't need it's “map code to POSIX errno” functionality, given we *are* throwing in the POSIX errno :)